src/builtin_string.cpp

```diff
@@ -1051,11 +1051,12 @@ static int string_repeat(parser_t &parser, io_streams_t &streams, int argc, wcha
 
     if ((to_repeat = string_get_arg(&i, argv, &storage, streams)) != NULL) {
         const wcstring word(to_repeat);
-        const wcstring full_string = max > 0 ? wcsrepeat_until(word, max) : wcsrepeat(word, count);
-        is_empty = full_string.empty();
+        const bool rep_until = (0 < max && word.length()*count > max) || !count;
+        const wcstring repeated = rep_until ? wcsrepeat_until(word, max) : wcsrepeat(word, count);
+        is_empty = repeated.empty();
 
         if (!quiet && !is_empty) {
-            streams.out.append(newline ? (full_string + L"\n") : full_string);
+            streams.out.append(newline ? (repeated + L"\n") : repeated);
```

Comment:
This is extremely inefficient as it will normally construct a new string due to the concatenation with a newline. It would be more efficient to just append `repeated`. Then append the newline character if that is needed.